File: 1/2020.09.28/77.Combinations/java/solution.java
class Solution {

    List<List<Integer>> res = new ArrayList<>();

    public List<List<Integer>> combine(int n, int k) {
        generate(1, new ArrayList<>(), n, k);
        return res;
    }

    private void generate(int start, List<Integer> tmpList, int n, int k) {
        // terminator
        if (k == 0) {
            res.add(new ArrayList<>(tmpList));
            return;
        }
        for (int i = start; i <= n - k + 1; i++) {
            // process current level
            tmpList.add(i);
            // drill down
            generate(i+1, tmpList, n, k - 1);
            // reset states
            tmpList.remove(tmpList.size() - 1);
        }
    }
}

File: 1/2020.09.28/77.Combinations/python/solution.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        def _generate(start, tmp, k):
            # terminator
            if k == 0:
                res.append(tmp[:])
                return
            for i in range(start, n - k + 2):
                # process current level
                tmp.append(i)
                # drill down
                _generate(i+1, tmp, k-1)
                # reset states
                tmp.pop()
        
        res = []
        _generate(1, [], k)
        return res
